tunnel-server/src/ssh/index.ts

```diff
@@ -22,48 +22,62 @@ export const createSshServer = ({
   tunnelUrl: (clientId: string, remotePath: string) => string
   helloBaseResponse: Record<string, unknown>
   tunnelsGauge: Pick<Gauge, 'inc' | 'dec'>
-}) => baseSshServer({
-  log,
-  sshPrivateKey,
-  socketDir,
-})
-  .on('client', client => {
-    const { clientId, publicKey, envId } = client
-    const pk = createPublicKey(publicKey.getPublicPEM())
+}) => {
+  const storeKeyToClient = new Map<string, BaseSshClient>()
+  const onClient = (client: BaseSshClient) => {
+    const { clientId, publicKey, envId, connectionId, publicKeyThumbprint, log } = client
     const tunnels = new Map<string, string>()
-    const jwkThumbprint = (async () => await calculateJwkThumbprintUri(await exportJWK(pk)))()
     client
       .on('forward', async (requestId, { path: tunnelPath, access, meta, inject }, localSocketPath, accept, reject) => {
         const key = activeTunnelStoreKey(clientId, tunnelPath)
+
         log.info('creating tunnel %s for localSocket %s', key, localSocketPath)
-        const setTx = await activeTunnelStore.set(key, {
+        const set = async (): ReturnType<typeof activeTunnelStore['set']> => await activeTunnelStore.set(key, {
           tunnelPath,
           envId,
           target: localSocketPath,
           clientId,
-          publicKey: pk,
+          publicKey,
           access,
           hostname: key,
-          publicKeyThumbprint: await jwkThumbprint,
+          publicKeyThumbprint,
           meta,
           inject,
-        }).catch(e => {
-          reject(
-            e instanceof KeyAlreadyExistsError
-              ? new Error(`duplicate path: ${key}, client map contains path: ${tunnels.has(key)}`)
-              : new Error(`error setting tunnel ${key}: ${e}`, { cause: e }),
-          )
+          client,
+        }).catch(async e => {
+          if (!(e instanceof KeyAlreadyExistsError)) {
+            throw e
+          }
+          const existingEntry = await activeTunnelStore.get(key)
+          if (!existingEntry) {
+            return await set() // retry
+          }
+          const otherClient = existingEntry.value.client as BaseSshClient
+          if (otherClient.connectionId === connectionId) {
+            throw new Error(`duplicate path: ${key}, from same connection ${connectionId}`)
```

Comment:
Does the handling need to be different in this case?

```
if (otherClient.connectionId === connectionId) {
            throw new Error(`duplicate path: ${key}, from same connection ${connectionId}`)
          }
```
If we log it and continue with the same flow, will we get to a correct end state?

Reply:
requesting a duplicate forward in the same client connection should reject the request in the `baseSshServer`, so this is handled as an edge case which is essentially a bug.